ethereum/contracts/ERC20App.sol

```diff
@@ -0,0 +1,124 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.5;
+pragma experimental ABIEncoderV2;
+
+import "@openzeppelin/contracts/access/AccessControl.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "./ScaleCodec.sol";
+import "./OutboundChannel.sol";
+
+enum ChannelId {Basic, Incentivized}
+
+contract ERC20App is AccessControl {
+    using ScaleCodec for uint256;
+
+    mapping(address => uint256) public balances;
+
+    mapping(ChannelId => Channel) public channels;
+
+    bytes2 constant MINT_CALL = 0x4201;
+
+    event Locked(
+        address token,
+        address sender,
+        bytes32 recipient,
+        uint256 amount
+    );
+
+    event Unlocked(
+        address token,
+        bytes32 sender,
+        address recipient,
+        uint256 amount
+    );
+
+    struct Channel {
+        address inbound;
+        address outbound;
+    }
+
+    bytes32 public constant INBOUND_CHANNEL_ROLE =
+        keccak256("INBOUND_CHANNEL_ROLE");
+
+    constructor(Channel memory _basic, Channel memory _incentivized) {
+        Channel storage c1 = channels[ChannelId.Basic];
+        c1.inbound = _basic.inbound;
+        c1.outbound = _basic.outbound;
+
+        Channel storage c2 = channels[ChannelId.Incentivized];
+        c2.inbound = _incentivized.inbound;
+        c2.outbound = _incentivized.outbound;
+
+        _setupRole(INBOUND_CHANNEL_ROLE, _basic.inbound);
+        _setupRole(INBOUND_CHANNEL_ROLE, _incentivized.inbound);
+    }
+
+    function lock(
+        address _token,
+        bytes32 _recipient,
+        uint256 _amount,
+        ChannelId _channelId
+    ) public {
+        require(
+            IERC20(_token).transferFrom(msg.sender, address(this), _amount),
```

Comment:
@philipstanislaus is there re-entrancy risk here? (eg: malicious erc20 that does weird shit in it's transferFrom)

Reply:
I'm guessing not, because even with re-entrancy I don't think anything can be stolen/fucked up? But should review in more detail.

Reply:
This should be fine, since there is no state written before the transferFrom call.

Reply:
will fix